src/calculi/mod.rs

```diff
@@ -0,0 +1,121 @@
+mod ert;
+use std::{fmt, rc::Rc};
+
+use ert::*;
+mod wp;
+use wp::*;
+mod wlp;
+use wlp::*;
+
+#[cfg(test)]
+mod tests;
+
+use crate::{
+    ast::{
+        visit::{walk_stmt, VisitorMut},
+        DeclKind, Direction, Files, Ident, SourceFilePath, Stmt, StmtKind,
+    },
+    intrinsic::annotations::{AnnotationError, AnnotationKind},
+    proof_rules::EncodingKind,
+    tyctx::TyCtx,
+};
+
+pub trait Calculus: fmt::Debug {
+    fn name(&self) -> Ident;
+
+    fn check_encoding(&self, encoding: EncodingKind, direction: Direction) -> Result<(), ()>;
```

Comment:
Instead of returning `()` in case of error, return a singleton error type, e.g. `CalculusAnnotationError` that implements `std::fmt::Error` (you can do this via the `thiserror` crate we're using).

Reply:
Also it would be good to add a comment describing what this method is supposed to do.

Reply:
Result<(),()> is used as a boolean success flag in this context. The returned error is wrapped with a proper `AnnotationError::CalculusEncodingMismatch` by the visitor. Should I use the singleton error type in this case?